### DISCUSSION TITLE
Parallel State Nodes Execution Order

### DISCUSSION THREAD

**Discussion by touchmarine** [statelyai/xstate#4143] [General]
Using parallel state nodes I would expect that transitions and actions are run in document order and the latter actions would receive updated context. For example, the following machine with initial context `x=false`, logs `A` and `B` on event `E`.

https://codesandbox.io/s/xstate-parallel-guards-tnfts6

<img width="240" alt="image" src="https://github.com/statelyai/xstate/assets/72550557/d4d6b4f1-867b-425b-9a9d-7467f79e1ddc">

Meanwhile, if I use `actions.choose` instead of guarded transitions, the machine logs only `A`.

https://codesandbox.io/s/xstate-parallel-actions-choose-guard-wlj63n

My question is if there is a reason that parallel states are not run in order and don't receive updated context? And, consequently, are `actions.choose` and `actions.pure` more than helpers?

**Comment by davidkpiano** [comment]
This is actually working as expected.

The guard (`cond`) on the transition determines whether a transition is **enabled**. So when the machine receives the `type: 'E'` event, it checks all regions in the parallel state to see which transitions are enabled _first_.

So on the first pass, if `isXFalse`, the transition will be marked as "enabled" _before_ any actions are called. So the result is that there are 2 enabled transitions: the one on `A0` and the one on `B0`. Therefore, both `A` and `B` will be logged. The queue of the actions looks something like this:

1. Log `A`
2. Assign `x: true`
3. Log `B`
4. Assign `x: true`

---

Now for the other example, both transitions are enabled because there are no guards, so all the actions will also be taken. However, note that now the actions _themselves_ are conditional due to `choose`. So the order will look like this:

1. If `isXFalse`: log `A` and assign `x: true`
2. ~~If `isXFalse`: log `B` and assign `x: true`~~
    - On this step, `x` is true, so `isXFalse` evaluates to false, because `x` was assigned to `true` in the previous step

Hope that clarifies things, and sorry if the intended behavior is confusing!

**Comment by touchmarine** [comment]
Thanks for the quick and detailed answer.

I did clone XState and played with it so I understand why the result is as is. I am wondering why the implementation is the way it is.

I checked the SCXML spec and didn't find anything concrete regarding processing of parallel states except the informative Algorithm for SCXML Interpretation. The following algorithm is, if I am correct, the same as XState's. But why is that? Is there a downside to executing actions and updating the context during the `_transition` method? Is it limiting something that I'm not seeing, perhaps static evalution?

And a bit as an aside regarding the ordering. Should parallel states be executed in document order? As I'm understanding the spec yes, but I believe you know the spec much better and am hoping to clarify this. The following machine should log `A` then `1` if it should be in document order.

https://codesandbox.io/s/xstate-parallel-statevalue-order-6p8qvx
